### ISSUE TITLE
Feature Request: .TGS format

### ISSUE THREAD

**Issue by GNUuser** [Wunderfitz/harbour-fernschreiber#100]
.TGS mean sticker animated.
I donate 20EU to Fernschreiber-Development for this LOVED feature to SailfishSO World. @jgibbon @monich @Wunderfitz

**Comment by GNUuser** [comment]
![clunmates](https://user-images.githubusercontent.com/45316884/96808024-fbed3f00-13e5-11eb-97b8-3d5ed41c09f2.jpg)

**Comment by Wunderfitz** [comment]
From my side, no donations needed, but TGS support is definitely on the list as most new sticker packages come in animated... However, no promises yet when and how it can be done...

**Comment by jgibbon** [comment]
TGS are gzipped Lottie-Animations which are basically "Bodymovin" animations (usually out of After Effects) in JSON format.

There are a few attempts to get lottie animations working in Qt, but none are actually feature complete, as far as I've read. Disclaimer: I haven't tried any of these.

 - An existing Qt labs [LottieAnimation](https://doc.qt.io/qt-5/qml-qt-labs-lottieqt-lottieanimation.html) which may or may not be possible to backport from [5.13](https://code.qt.io/cgit/qt/qtlottie.git/log/?h=5.13) or [later](https://code.qt.io/cgit/qt/qtlottie.git/). Thankfully, Fernschreiber is GPL3, but there were some other license considerations for including qt modules, weren't there? This _may_ be the best way and could possibly be extended to gunzip the data "in place". But I think we'd need someone with pretty decent c++ knowledge to even realistically evaluate this as a possible option, let alone backport/implement it. @monich would you be up for such shenanigans? ;)
 - [Lottie QML](https://github.com/kbroulik/lottie-qml) which uses/wraps the js player for usage in QML/Canvas. I don't think this can possibly be very fast. Development does not seem active but there's a fork on github that at least tried to fix a memory issue.
 - [Lottie QtQuick](https://github.com/Conntac/lottie-qt) which does seem pretty experimental from first glance and isn't actively developed.

cheers

**Comment by monich** [comment]
> An existing Qt labs [LottieAnimation](https://doc.qt.io/qt-5/qml-qt-labs-lottieqt-lottieanimation.html) which may or may not be possible to backport from [5.13](https://code.qt.io/cgit/qt/qtlottie.git/log/?h=5.13) or [later](https://code.qt.io/cgit/qt/qtlottie.git/). Thankfully, Fernschreiber is GPL3, but there were some other license considerations for including qt modules, weren't there? This _may_ be the best way and could possibly be extended to gunzip the data "in place". But I think we'd need someone with pretty decent c++ knowledge to even realistically evaluate this as a possible option, let alone backport/implement it. @monich would you be up for such shenanigans? ;)

I was able to compile it against Qt 5.6 with very little patching. Seems to be doable from the technical PoV,  no idea about licensing.

**Comment by jgibbon** [comment]
It indeed seems to be a monster. But that sounds like good news! At least if the implemented animation subset is enough for stickers, which we'll only be able to tell after it's done. 

After checking a bit more, I think licensing should be a non issue with the GPL3 module, provided it's attributed and all changes are also made public – the usual. I may have had some old "do not link modules statically" rule or something in the back of my head. https://code.qt.io/cgit/qt/qtlottie.git/tree/LICENSE.GPL3-EXCEPT

I guess the "right" way for Linux in general would be to package it separately, but in the other hand it may also be good to integrate it into the package like tdlib in this case. Or even ditch the module part and integrate it into the compiled binary?

I've never really done either (or at least done it right) and couldn't really say how to do it.

**Comment by monich** [comment]
It can be integrated as a binary the same way as `libtdjson.so.1.6.0` is done (with `RPATH` pointing to `/usr/share/harbour-fernschreiber/lib`) there's no need to compile it in statically.

**Comment by Wunderfitz** [comment]
Mhmm, that sounds like a plan. Wow, really looking forward - @monich have you already tried it out or was it just the compilation effort for now?

Unzipping the TGS file should be rather straight forward. libz is there (and even Harbour-compliant) and there is [Quazip](https://stachenov.github.io/quazip/) which can easily be integrated as well...

**Comment by monich** [comment]
> Mhmm, that sounds like a plan. Wow, really looking forward - @monich have you already tried it out or was it just the compilation effort for now?

[This](https://github.com/monich/qtlottie) builds in Mer OBS against Sailfish OS 3.0.0.8 (that's Qt 5.6.3), with some warnings that don't seem critical, but I haven't tried to actually run it. I doubt that I can do much more during the week.

**Comment by monich** [comment]
Installed and ran [qtlottie](https://code.qt.io/cgit/qt/qtlottie.git/) test on a device, and unfortunately it doesn't work with (uncompressed) tgs files :/
There are a few warnings like this:
```
[W] BMShape::construct:163 - Unsupported shape type: "mm"
[W] BMShape::construct:163 - Unsupported shape type: "mm"
[W] BMLayer::construct:87 - Unsupported layer type: 3
```
although `LottieAnimation` seems to parse JSON more or less all right, sizes itself to the actual dimensions of the sticker animation, rendering thread is running but nothing is being drawn on the screen (sometimes a flashing dot). It's a false start, it seems.

**Comment by Wunderfitz** [comment]
Thank you very much for trying it out! Though it slightly disappointing (hope it wasn't too time-consuming...) I'm sure there will be an option that works somewhen...

**Comment by jgibbon** [comment]
Tdesktop (also Qt) somehow uses liblottie directly: https://github.com/telegramdesktop/tdesktop/search?q=lottie

**Comment by monich** [comment]
> Tdesktop (also Qt) somehow uses liblottie directly: https://github.com/telegramdesktop/tdesktop/search?q=lottie

That's coming from [here](https://github.com/desktop-app/lib_lottie), I'll take a look at that one. It Tdesktop is using it, it must work with stickers.

**Comment by monich** [comment]
AFAICT it's actually [rlottie](https://github.com/Samsung/rlottie) which does tgs rendering for tdesktop, that `lib_lottie` thing seems to be just a wrapper which pulls in tons of additional dependencies.

**Comment by Wunderfitz** [comment]
I had some time this afternoon to experiment a bit. rlottie compiles smoothly in the latest Sailfish OS SDK and produces a nice .so file that could potentially be used - if not directly then at least to produce an animated GIF that can be used natively. But I don't have time in the next days to continue there unfortunately.

**Comment by monich** [comment]
> I had some time this afternoon to experiment a bit. rlottie compiles smoothly in the latest Sailfish OS SDK and produces a nice .so file that could potentially be used - if not directly then at least to produce an animated GIF that can be used natively. But I don't have time in the next days to continue there unfortunately.

To me, a pretty obvious approach would be to write a TGS [QImageIOHandler](https://doc.qt.io/qt-5/qimageiohandler.html) using rlottie for rendering. I'm quite surprized that I haven't found such a component. Perhaps I should write one, then.

**Comment by monich** [comment]
> ...
> To me, a pretty obvious approach would be to write a TGS [QImageIOHandler](https://doc.qt.io/qt-5/qimageiohandler.html) using rlottie for rendering. I'm quite surprized that I haven't found such a component. Perhaps I should write one, then.

It wasn't that hard, indeed. And even works quite nicely. The last remaining serious issue is that synchronous rendering produces uneven frame rate, and if I enable built-in async rendering in rlottie, it behaves significantly better for a while but eventually crashes like this:
```
terminate called after throwing an instance of 'std::future_error'
  what():  std::future_error: Promise already satisfied

Thread 18 "harbour-fernsch" received signal SIGABRT, Aborted.
[Switching to Thread 0xe7bd0230 (LWP 13243)]
__libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
47		pop	{r7, pc}
(gdb) bt
#0  __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
#1  0xf5dc6bda in __libc_signal_restore_set (set=0xe7bcf82c) at ../sysdeps/unix/sysv/linux/internal-signals.h:84
#2  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:48
#3  0xf5db07b2 in __GI_abort () at abort.c:79
#4  0xf5fabf70 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#5  0xf5faa8e4 in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:47
#6  0xf5faa93a in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:57
#7  0xf5faab72 in __cxxabiv1::__cxa_throw (obj=obj@entry=0xe566d8, tinfo=0xf603d3d4 <typeinfo for std::future_error>, dest=0xf5fc5619 <std::future_error::~future_error()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:95
#8  0xf5fc58ec in std::__throw_future_error (__i=2) at /usr/src/debug/gcc-8.3.0-1.3.58.jolla.arm/obj-armv7hl-meego-linux-gnueabi/armv7hl-meego-linux-gnueabi/libstdc++-v3/include/system_error:151
#9  0x0056434c in std::__future_base::_State_baseV2::_M_set_result(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>, bool) (this=0xe8f7945c, __res=..., __ignore_failure=false)
    at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/future:408
#10 0x00565d9e in std::promise<rlottie::Surface>::set_value (this=0xc9ff8c, __r=...) at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/future:1110
#11 0x00564cba in RenderTaskScheduler::run (this=0x6a4e74 <RenderTaskScheduler::instance()::singleton>, i=1) at ../../../rlottie/src/lottie/lottieanimation.cpp:172
#12 0x00564cfa in RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}::operator()() const (__closure=0xe9e134) at ../../../rlottie/src/lottie/lottieanimation.cpp:179
#13 0x00569a88 in std::__invoke_impl<void, RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}>(std::__invoke_other, RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}&&) (__f=...)
    at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/bits/invoke.h:60
#14 0x00569006 in std::__invoke<RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}>(std::__invoke_result&&, (RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}&&)...) (__fn=...)
    at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/bits/invoke.h:95
#15 0x0056bcfa in std::thread::_Invoker<std::tuple<RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}> >::_M_invoke<0u>(std::_Index_tuple<0u>) (this=0xe9e134) at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/thread:244
#16 0x0056bc4e in std::thread::_Invoker<std::tuple<RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}> >::operator()() (this=0xe9e134) at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/thread:253
#17 0x0056ba42 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<RenderTaskScheduler::RenderTaskScheduler()::{lambda()#1}> > >::_M_run() (this=0xe9e130) at /opt/cross/armv7hl-meego-linux-gnueabi/include/c++/8.3.0/thread:196
#18 0xf5fc6ad6 in std::execute_native_thread_routine (__p=0xe9e130) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#19 0xf604ae6a in start_thread (arg=0x0) at pthread_create.c:479
#20 0xf5e4c21c in ?? () at ../sysdeps/unix/sysv/linux/arm/clone.S:73 from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) inf thr
  Id   Target Id                                       Frame 
  1    Thread 0xf348e760 (LWP 12400) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  2    Thread 0xf0ec5230 (LWP 12493) "QQmlThread"      __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  3    Thread 0xf02ff230 (LWP 12500) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
  4    Thread 0xf00fe230 (LWP 12501) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
  5    Thread 0xefcff230 (LWP 12502) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
  6    Thread 0xef8ff230 (LWP 12503) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
  7    Thread 0xef1fe230 (LWP 12504) "QDBusConnection" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  8    Thread 0xeeffd230 (LWP 12505) "TDLibReceiver"   __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  9    Thread 0xee7fb230 (LWP 12526) "gmain"           __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  10   Thread 0xee5fa230 (LWP 12527) "dconf worker"    __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  11   Thread 0xee3f9230 (LWP 12528) "gdbus"           __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  12   Thread 0xeddff230 (LWP 12535) "QSGRenderThread" 0xf6cc0bea in QSGAtlasTexture::swizzleBGRAToRGBA (image=0xeddfe570) at scenegraph/util/qsgatlastexture.cpp:226
  13   Thread 0xedbfe230 (LWP 12542) "NemoThumbnailLo" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  14   Thread 0xed9fd230 (LWP 12543) "QSGRenderThread" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  15   Thread 0xed3ff230 (LWP 12544) "QQuickPixmapRea" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  16   Thread 0xe872f230 (LWP 13228) "NemoThumbnailLo" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:46
  17   Thread 0xe7dd1230 (LWP 13242) "harbour-fernsch" 0x005bdace in VPoint::x (this=0xe7dd0a54) at ../../../rlottie/src/vector/vpoint.h:114
* 18   Thread 0xe7bd0230 (LWP 13243) "harbour-fernsch" __libc_do_syscall () at ../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47
  19   Thread 0xe79cf230 (LWP 13244) "harbour-fernsch" gray_hline (worker=0xe79cdce8, x=378, y=183, area=90624, acount=1) at ../../../rlottie/src/vector/freetype/v_ft_raster.cpp:894
  20   Thread 0xe77ce230 (LWP 13245) "harbour-fernsch" ft_trig_pseudo_rotate (vec=0xe77cd898, theta=33809) at ../../../rlottie/src/vector/freetype/v_ft_math.cpp:209
(gdb) 
```
as if something isn't getting properly synchronized. Once I resolve that (either by fixing the bug causing the panic or by disabling rlottie threading and switching to Qt-provided thread pooling which works reliably) I'll submit a pull request. Stay tuned.

**Comment by monich** [comment]
Yey, I think I found a bug in my code. It's my lack of practice in using the standard C++ template/class library. Now it's been animating several stickers in parallel for an hour or so without any crashes. Almost there))

**Comment by jgibbon** [comment]
Sounds like amazing work, impressive! I'm looking forward to checking that out.

That reminds me: 
As a few users have mentioned their strong distaste for (animated) stickers, we probably should consider a setting to "opt out" of animations before releasing something. That doesn't mean you should implement/include that for your upcoming PR, only that we may think about and possibly add it before the following release.

**Comment by jgibbon** [comment]
Also, we have to comply with the collected licenses of https://github.com/Samsung/rlottie/blob/master/COPYING
@Wunderfitz do you think "mentioning" it at the appropriate places may be enough? 

Sorry for spamming this issue with my collected thoughts one at a time. ;)

**Comment by monich** [comment]
> As a few users have mentioned their strong distaste for (animated) stickers, we probably should consider a setting to "opt out" of animations before releasing something. That doesn't mean you should implement/include that for your upcoming PR, only that we may think about and possibly add it before the following release.

It's definitely going to be configurable, if not for aesthetic then for performance reasons. Caching all frames eats up tons of memory (each frame is 512x512x4=1MB, animations typically contains from 30 to 90 frames, several animations in one chat can easily take hundreds MB of heap), I ended up constantly re-rendering them, which eats up all the available CPU time (still missing some frames even on a pretty powerful XA2 esp. when several animations are running in parallel) and drains the battery. The option to turn it off is a must.

**Comment by Wunderfitz** [comment]
Wow, these are really awesome news. In this case, I would wait with the 0.4 release until we've integrated this PR. Would be ridiculous to release a new version if such a feature is around the corner.

Maybe it would be one option to do it the same way as with it's already done for videos and animated GIFs. A preview image is loaded in any case and the user can play the animation explicitly. But some people may also want to play everything or nothing...

On the licensing front, I'm pretty relaxed. It seems that (except one) all licenses are compatible with GPLv3. The only exception is the JSON license which says "The Software shall be used for Good, not Evil." The FSF says that this contradicts freedom 0 and is therefore non-free, but I would use it anyway as a good Telegram client for Sailfish OS is always for Good! 😄 We will need to mention rlottie on the About page (along with TDLib and Twemoji), but if that lib works, I would be glad to mention them.

**Comment by monich** [comment]
To me, manual control of playing GIFs and stickers makes no sense at all. They either play or they don't. In other words, they should be treated as animated images, rather than videos. Animated images are smaller and shorter than videos, it's ok to load them automatically, they are displayed inline rather that full screen, they don't need positioning, fast forwarding or anything like that, and so on.

And I would say GIFs and stickers should have different config options for animation because a) stickers are going to be heavier, it seems and b) those are two slightly different use cases. Animated stickers are typically designed so that they still make sense even if they are not animated. Animated GIFs on the other hand only make sense as an animation, most of them are pretty useless as a static image.

**Comment by jgibbon** [comment]
Manual playback _can_ have potential advantages for both gif and stickers, depending on the user... 
Starting with pure performance/energy reasons with the ability to see the whole thing if the user explicitly decides to. Especially with that one person everyone has in some group who spams about 100 dank gif memes per hour. 
Also I like to decide when a video/gif begins and not start watching the end. 

Both should also pause with the application in background, I think.

**Comment by monich** [comment]
> Manual playback can have potential advantages for both gif and stickers, depending on the user...
Starting with pure performance/energy reasons with the ability to see the whole thing if the user explicitly decides to. Especially with that one person everyone has in some group who spams about 100 dank gif memes per hour.
Also I like to decide when a video/gif begins and not start watching the end.

Yeah, it's not an easy task to make animated GIF elements convenient, flexible and visually pleasing at the same time. Ability to stop/pause the animation and restart it from the beginning do sound like useful features to me. But I wouldn't want to see a bunch of controls on top of every animated image which is not currently being animated. On the other hand, how to tell then which ones are animated and which are not. I'm not sure. ¯\_(ツ)_/¯

And as I already explained, animated images and videos are different things IMO. Although the distinction is somewhat blurry.

> Both should also pause with the application in background, I think.

That goes without saying

**Comment by monich** [comment]
[This](https://github.com/Wunderfitz/harbour-fernschreiber/pull/140) is what I came up with. TGS only, haven't touched GIFs or anything else.

**Comment by monich** [comment]
Done, I guess?

**Comment by Wunderfitz** [comment]
Yep, that's done. Thanks again for that awesome contribution, @monich!

**Comment by GNUuser** [comment]
Wow Wow Wowwww. 
A HUGE Thanks to @monich @jgibbon @Wunderfitz

**Comment by GNUuser** [comment]
YottaGram is  cloned Fernschreiber-features-like: TGS  
But Fernschreiber will be forever  the first one in support and created. :  )

**Comment by Wunderfitz** [comment]
Well, that's the beauty of open source. Moreover, fair competition is really good - users will get cool Telegram clients and in the end the whole platform benefits - that's what we all want in the end. :) In this case, credits go entirely to @monich - he implemented that! :+1:
